apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RpcRetryConsumer.cs

```diff
@@ -0,0 +1,209 @@
+    /// An <see cref="IConsumer{T}"/> that will allow for the option to retry RPC failures and buffer
+    /// the data while the retries occur.
+    /// </summary>
+    internal class RpcRetryConsumer<T> : IConsumer<T>
+    {
+        /// <summary>The options dictating how retries will be handled.</summary>
+        private readonly RetryOptions _options;
+
+        /// <summary>A function to obtain the size of an item in bytes.</summary>
+        private readonly Func<T, int> _sizer;
+
+        /// <summary>
+        /// Function to create a <see cref="ISequentialThreadingTimer"/> that will start the given
+        /// action based on retry options.
+        /// </summary>
+        private readonly Func<Action, RetryOptions, ISequentialThreadingTimer> _timerFactory;
+
+        /// <summary>The timer running in the background to retry failed requests.</summary>
+        private readonly ISequentialThreadingTimer _timer;
+
+        /// <summary>An mutex to protect the buffer.</summary>
+        private readonly object _mutex = new object();
+
+        /// <summary>The consumer to send items to.</summary>
+        internal readonly IConsumer<T> _consumer;
+
+        /// <summary>
+        /// The buffered items. This is not readonly as it is replaced when the buffer is successfully send to
+        /// the base consumer.
+        /// </summary>
+        private List<T> _buffer = new List<T>();
+
+        /// <summary>The size of the buffer.</summary>
+        private int _sizeInBytes = 0;
+
+        /// <summary>The number of sequential unsuccessful retries.</summary>
+        private int _retries = 0;        
+
+        internal RpcRetryConsumer(IConsumer<T> consumer, RetryOptions options, 
+            Func<T, int> sizer, Func<Action, RetryOptions, ISequentialThreadingTimer> timerFactory)
+        {
+            _consumer = GaxPreconditions.CheckNotNull(consumer, nameof(consumer));
+            _options = GaxPreconditions.CheckNotNull(options, nameof(options));
+            _sizer = GaxPreconditions.CheckNotNull(sizer, nameof(sizer));
+            _timerFactory = GaxPreconditions.CheckNotNull(timerFactory, nameof(timerFactory));
+
+            // Start the background timer that will retry failed requests.
+            _timer = _timerFactory(FlushBuffer, _options);
+        }
+
+        /// <inheritdoc />
+        public void Receive(IEnumerable<T> items)
+        {
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                _consumer.Receive(items);
+            }
+            catch (RpcException e) when (HandleException(items, e)) { }
+        }
+
+        /// <inheritdoc />
+        public Task ReceiveAsync(IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                return _consumer.ReceiveAsync(items, cancellationToken);
```

Comment:
One aspect about this that might be a little odd. Suppose we have two entries, E1 and E2. The first fails, so needs to be retried, but the second succeeds. We could easily end up sending E2 before E1, because E1 is only sent on a timer. Is that a problem?